### ISSUE TITLE
Request for Input: Allow IBlockchainProcessingOrchestrator to checkpoint IBlockProgressRepository on BlockRangeRequestStrategy

### ISSUE THREAD

**Issue by jamescarter-le** [Nethereum/Nethereum#742]
I have a scenario with a long running BlockchainProcessor.

When I want to run this to catch up with older points in time, where either processing time is long enough that a failure would loose significant processing time, or where the memory space used by the aggregation process exceeds the number of blocks from BlockFrom - CurrentBlock, the `UpdateLastBlockProcessedAsync` is only processed once we ready the Last Confirmed block number.

I would like to change this so that `UpdateLastBlockProcessedAsync` is invoked every time `BlockRangeRequestStrategy` limit is reached.

In my example, I am processing swap data from Uniswap V3 pools. I want to checkpoint the data every ~24hrs. So I would pass a `BlockRangeRequestStrategy` with `NumberOfBlocksPerRequest` of 6000. This means on a failure I only loose 6000 blocks of processing, rather than the whole data set.

When `UpdateLastBlockProcessedAsync` is invoked, I can persist the data constructed in my unit of work to the DB, along with the block number.

I have added the following `IBlockProgressRepository` so I can persist units of work:

```
    public class FuncBlockProgressRepository : IBlockProgressRepository
    {
        private readonly Func<Task<BigInteger?>> getLastBlockNumberFunc;
        private readonly Func<BigInteger, Task> upsertProgressAsyncFunc;

        public FuncBlockProgressRepository(Func<Task<BigInteger?>> getLastBlockNumberFunc, Func<BigInteger, Task> upsertProgressAsyncFunc)
        {
            if (getLastBlockNumberFunc == null)
                throw new ArgumentNullException(nameof(getLastBlockNumberFunc));

            if (upsertProgressAsyncFunc == null)
                throw new ArgumentNullException(nameof(upsertProgressAsyncFunc));

            this.getLastBlockNumberFunc = getLastBlockNumberFunc;
            this.upsertProgressAsyncFunc = upsertProgressAsyncFunc;
        }

        public Task<BigInteger?> GetLastBlockNumberProcessedAsync()
        {
            return getLastBlockNumberFunc();
        }

        public Task UpsertProgressAsync(BigInteger blockNumber)
        {
            return upsertProgressAsyncFunc(blockNumber);
        }
    }
```

However I would also want to change `LogOrchestrator` as currently it runs up through From - To, split by the `_blockRangeRequestStrategy`. What I would like, is for this to return `OrchestrationProgress` as soon as it processes a single block of logs, so that `BlockchainProcessor` can invoke `UpdateLastBlockProcessedAsync`. For this I suggest moving `BlockRangeRequestStrategy` up into `BlockchainProcessor`, and dropping this from  `LogOrchestrator`.

Happy to take feedback on this for the best way forward.

TLDR: I want `_blockProgressRepository.UpsertProgressAsync` to be invoked more frequently when processing Batches, than when we have reached the Current block. When progressing from Block 0 to Current, we will only receive one opportunity to persist aggregation.

**Comment by juanfranblanco** [comment]
A failure should not cause to lose progress, as these are handled by the Processor, the idea to not track the progress directly on the orchestrator was to avoid costs of writing those records constantly, one record at time when crawling.

BUT! This does not count for a catastrophic error, mainly a machine going down, a very simple scenario will be to have a spot instance to do some crawling / log processing for storage that can go down any time. So yeah it is a good idea to track it in case of not software error as it can be running for days and the machine shutdown.

The simplest solution, as the circumstances are completely different from crawling and log processing (ie not moving the BlockRangeRequestStrategy) is to have each orchestrator do the update of the progress. 

For compatibility (not breaking changes) the best solution is to add the progress repository as a parameter to the orchestrators, and do the update per block or range depending of the scenario.

Ill add that..

**Comment by juanfranblanco** [comment]
Cool thanks, just to make sure the progress repository only updates a single record of the current block number (default ones), although yeah it can be tracked or even create your own custom, send a notification.. etc.

**Comment by juanfranblanco** [comment]
This is included in the following commit: https://github.com/Nethereum/Nethereum/commit/1f52f0b968b4f315a89ad3941d63aea872b6f9e4

**Comment by jamescarter-le** [comment]
Amazing! Thanks @juanfranblanco
